🚘 Auto opt cli

trajepl opened the pull request:
## Describe your changes

Auto opt cli.

e.g. for bert model, we can optimize the model from huggingface and conduct onnx model with:
```sh

olive auto-opt --model Intel/bert-base-uncased-mrpc --data_config_path data_config.json --task text-classification

olive auto-opt --model Intel/bert-base-uncased-mrpc --data_config_path data_config.json --task text-classification --precision int4 --providers CPU

olive auto-opt --model Intel/bert-base-uncased-mrpc --data_config_path data_config.json --task text-classification --precision fp16 --providers CUDA

# use model builder
olive auto-opt --model microsoft/Phi-3-mini-4k-instruct --precision fp16 --providers CUDA --use_model_builder

```

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

## (Optional) Issue link

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+            help="The task of the input model. Need to be specified if the input model is downloaded from huggingface.",
+        )
+        model_group.add_argument(
+            "--model_type",
```
Let's avoid overloading "model_type". May be "model_framework"? or "model_kind"? or something else?

trajepl replied:
adopted `model_framework`

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+            "--device",
+            type=str,
+            default=None,
+            choices=["cpu", "gpu"],
```
Add "npu" also

trajepl replied:
Updated.

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+            "--providers",
+            type=str,
+            nargs="*",
+            choices=["CPU", "CUDA", "Tensorrt", "Dml", "VitisAI"],
```
QnnEP also

trajepl replied:
Updated.

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+        )
+        search_strategy_group.add_argument("--seed", type=int, default=0, help="Random seed for search algorithm")
+        search_strategy_group.add_argument(
+            "--execution_order",
```
"search_order" is clearer term here.

trajepl replied:
Updated.

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+            help="Execution order for search strategy",
+        )
+        search_strategy_group.add_argument(
+            "--search_algorithm",
```
We want to disable search, evaluation by default. Which means. dataset options should be optional.

trajepl replied:
Updated.

devang-ml commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,230 @@
+
+        search_strategy_group = sub_parser.add_argument_group("search strategy options")
+        search_strategy_group.add_argument(
+            "--num-samples", type=int, default=5, help="Number of samples for search algorithm"
```
Do we really need to expose this in the CLI?

trajepl replied:
I was thinking if the search takes long time, user can somehow reduce the numbers of num-samples to stop the search in time.

devang-ml reviewed (commented):
Let's add "auto-opt" needed packages list under extra dependencies so that user can do `pip install olive-ai[auto-opt]`

trajepl commented:
> Let's add "auto-opt" needed packages list under extra dependencies so that user can do `pip install olive-ai[auto-opt]`

I think it might be hard to define a unified extra dependencies lists for auto-opt. Since auto-opt may be used in different device which cause the conflicts on onnxruntime version.

I think we have a feature to dynamicly `get_required_packages` based on the given accelerators which can be used in auto-opt.

devang-ml commented:
We can exclude OnnxRuntime and IHV toolkits. But we should be able to install other required packages using `pip install olive-ai[auto-opt]`.

trajepl commented:
> We can exclude OnnxRuntime and IHV toolkits. But we should be able to install other required packages using `pip install olive-ai[auto-opt]`.

Updated. Basically, `pip install olive-ai` is adequate to run auto-opt cli. Furthermore, to simplify the conversion, since olive currently use `optimum` to get onnx model, I added `optimum` to olive-ai[auto-opt]. 

Also I tested that, after the installation olive-ai[auto-opt], user only needs to install corresponding onnxruntime or onnxruntime-genai(_only enable --use_model_builder_) based on the device(cpu/gpu etc.), olive auto-opt cli can conduct reasonable results.

samuel100 commented:
What models will be supported by `auto-opt`? When I have tried the auto optimizer using:

```json
{
    "input_model":{
        "type": "HfModel",
        "model_path": "microsoft/phi-3.5-mini-instruct",
        "task": "text-generation"
    },
    "systems": {
        "local_system": {
            "type": "LocalSystem",
            "accelerators": [
                {
                    "device": "cpu",
                    "execution_providers": [
                        "CPUExecutionProvider"
                    ]
                }
            ]
        }
    },
    "auto_optimizer_config": {
        "opt_level": 0,
        "disable_auto_optimizer": false,
        "precision": "int4"
    },
    "host": "local_system",
    "target": "local_system",
    "cache_dir": "cache",
    "output_dir" : "models"
}
```
I get an error message on the OrtTransformerOptimizer pass:

```
ValueError: Unsupported model type: phi3, please select one from [bart, bert, bert_tf, bert_keras, clip, gpt2, gpt2_tf, gpt_neox, swin, tnlr, t5, unet, vae, vit, conformer, phi] which need to be set under OrtTransformersOptimization.config
```

If these are the only models supported then it is a little underwhelming because they are pretty outdated. It is also a bit odd because the optimizer works for optimum models when I run `olive finetune`. 

As a general rule, a user should be able to plug in:

- A wide range of model architectures (for example, if Llama4 is released and it has the same architecture as Llama3, then I'd expect the tool to work).
- Device target
- precision

The user should not need to worry about adding data.

devang-ml commented:
The list and the error message is from OnnxRuntime's transformer optimizer.

devang-ml commented:
Model supported by the transformer optimizer are listed at https://github.com/microsoft/onnxruntime/blob/f4d62eeb2e058e2c7b5de0eaa9599368d32b23d5/onnxruntime/python/tools/transformers/optimizer.py#L50

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+        )
+
+        # model options
+        model_group = sub_parser.add_argument_group("model options")
```
Can you use [this](https://github.com/microsoft/Olive/blob/main/olive/cli/base.py#L105) and [this](https://github.com/microsoft/Olive/blob/main/olive/cli/base.py#L42) method here?

trajepl replied:
Done.

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+
+        # output options
+        output_group = sub_parser.add_argument_group("output options")
+        output_group.add_argument(
```
Not related to this PR but we could gather this to base.py as well.

trajepl replied:
Good call. But I saw there are different arguments attributes. Some of them are required, some of them are not. And they have different default values. 

Can we update it in follow-up PR?

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+                # TODO(jambayk): point user to datastore with outputs or download outputs
+                # both are not implemented yet
+                return
+            output_path = Path(self.args.output_path)
```
I updated the cli output logic for other clis. Can you take a look?

trajepl replied:
Done.

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+            device = (
+                "gpu"
+                if self.args.providers
+                and any(p[: -(len("ExecutionProvider"))] in ["CUDA", "Tensorrt", "Dml"] for p in self.args.providers)
```
```
        system_group.add_argument(
            "--providers",
            type=str,
            nargs="*",
            choices=["CPU", "CUDA", "Tensorrt", "Dml", "VitisAI", "Qnn"],
            help="List of execution providers to use for optimization",
        )
```
without `ExecutionProvider`?

xiaoyu-work replied:
Should device be set to cpu if `VitisAI` and `Qnn` provided here?

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+                and any(p[: -(len("ExecutionProvider"))] in ["CUDA", "Tensorrt", "Dml"] for p in self.args.providers)
+                else "cpu"
+            )
+        providers = self.args.providers or ["CPUExecutionProvider"] if device == "cpu" else ["CUDAExecutionProvider"]
```
Same as above

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+            "num_samples": self.args.num_samples,
+            "seed": self.args.seed,
+        }
+        # TODO(anyone): add remote options to auto opt passes
```
Is remote option not working here?

xiaoyu-work commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,255 @@
+    },
+    "host": "local_system",
+    "evaluators": EVALUATE_TEMPLATE,
+    "evaluator": "common_evaluator",
```
What does this `evaluator` do?

samuel100 commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,265 @@
+            ),
+        )
+        auto_opt_config_group.add_argument(
+            "--use_model_builder",
```
Can we abstract MB decision point for user? As a first-time user, this will add cognitive load ("what is model builder; should I use this"). Instead, could we test whether MB supports the model and if it does just use MB?

![image](https://github.com/user-attachments/assets/8dbf9835-fe58-4390-b476-25c8287404c8)

devang-ml replied:
We need an API from MB to check whether the model is supported or not. We can do this as a follow up PR.

samuel100 commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,265 @@
+            "dataset options, required for some optimization passes like quantization, and evaluation components"
+        )
+        dataset_group.add_argument(
+            "--data_config_path",
```
Could we make this a path to HF/Local to the actual data without requiring additional configuration? What does the data need to look like, how would the user generate it?

devang-ml replied:
We need to do the same as 'finetune` command and avoid having user provide json/yaml config file.

samuel100 commented on olive/cli/auto_opt.py:
```diff
@@ -0,0 +1,265 @@
+    @staticmethod
+    def register_subcommand(parser: ArgumentParser):
+        sub_parser = parser.add_parser(
+            "auto-opt",
```
Should we simplify the name of the command to `optimize`?
